src/ocl/convolutionocl.cpp

```diff
@@ -655,48 +670,62 @@ static void DirConvFindCore(Handle& handle,
+                handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
+            const auto implictgemm_solutions = conv.FindDataImplicitGemmSolutions(
+                handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
+            const auto fft_solutions = conv.FindFftSolutions(ctx, invoke_ctx);
+            all_solutions.insert(
+                all_solutions.end(), direct_solutions.begin(), direct_solutions.end());
+            all_solutions.insert(
+                all_solutions.end(), implictgemm_solutions.begin(), implictgemm_solutions.end());
+            all_solutions.insert(all_solutions.end(), fft_solutions.begin(), fft_solutions.end());
+        }
     }
 
-    // Implicit GEMM algo
-    if(!use_winograd_only)
+    if(!all_solutions.empty())
     {
-        ConvolutionUserBuffers bufs(workSpace, workSpaceSize);
-        bufs.SetFwd(x, w, y);
-        const auto all = conv.FindDataImplicitGemmSolutions(
-            handle, xDesc, wDesc, yDesc, exhaustiveSearch, true, bufs, invoke_ctx);
-        PrecompileSolutions(handle, all);
-        const auto algorithm_name = AlgorithmName{"miopenConvolutionFwdAlgoImplicitGEMM"};
-        EvaluateInvokers(handle, all, algorithm_name, network_config, invoke_ctx, record);
-    }
+        // Precompile Solutions
+        PrecompileSolutions(handle, all_solutions);
 
-    // FFT algo
-    if(!use_winograd_only && !miopen::IsDisabled(MIOPEN_DEBUG_CONV_FFT{}))
-    {
-        const auto all            = FindAllFFTSolutions(ctx, invoke_ctx);
-        const auto algorithm_name = AlgorithmName{"miopenConvolutionFwdAlgoFFT"};
-        PrecompileSolutions(handle, all);
-        EvaluateInvokers(handle, all, algorithm_name, network_config, invoke_ctx, record);
+        // Evaluate Invokers
+        auto iter_start = all_solutions.begin();
+        auto iter       = all_solutions.begin();
+        std::string algo_name =
+            solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
+        for(; iter != all_solutions.end(); ++iter)
+        {
+            if(solver::Id(iter->solver_id).GetAlgo(ctx.conv_problem.GetDirection()) != algo_name)
+            {
+                EvaluateInvokers(handle,
+                                 {iter_start, iter},
+                                 AlgorithmName(algo_name),
+                                 network_config,
+                                 invoke_ctx,
+                                 record);
+                iter_start = iter;
+                algo_name =
+                    solver::Id(iter_start->solver_id).GetAlgo(ctx.conv_problem.GetDirection());
+            }
+        }
+        EvaluateInvokers(handle,
+                         {iter_start, iter},
+                         AlgorithmName(algo_name),
+                         network_config,
+                         invoke_ctx,
+                         record);
```

Comment:
[Recommendation] This code assumes that input vector is ordered, so that solutions of the same algorithm belong to the same range. And AFAICS you are trying to avoid copying of vectors. However the code still looks ugly (or tricky at least). Would you like to rework it? I can throw a couple of ideas if you wish.

Reply:
Yes, I'm very glad to be able to improve my skills and contribute to `miopen`.

Reply:
Nice to hear that ;) Recommendation:
```C++
...
const auto direct = conv.FindDataDirectSolutions(...);
...
const auto igemm = conv.FindDataImplicitGemmSolutions(...);
...
{
    std::vector<solver::ConvSolution*> all; // pointers
    for(&s: direct) all.emplace_back(&s);
    for(&s: igemm) all.emplace_back(&s);
    ...
    // Make PrecompileSolutions accept std::vector<solver::ConvSolution*>&, then:
    PrecompileSolutions(handle, all);
}
EvaluateInvokers(handle, direct, "miopenConvolutionFwdAlgoDirect" ...);
EvaluateInvokers(handle, igemm, "miopenConvolutionFwdAlgoImplicitGEMM" ...);
...
```
I guess we just eliminated two unnecessary copies of Solutions (the first happens when the result of `FindDataDirectSolutions()` etc is copied into `all_solutions`, the second - during `EvaluateInvokers(... {iter_start, iter} ... )`, as you already know) and nontrivial loop.